### PR TITLE
feat: diffusion LLM abliteration support (DiffusionGemma)

### LINKED ISSUE #370: Diffusion Models abliteration :)
A new challenge has arrived

https://huggingface.co/google/diffusiongemma-26B-A4B-it

### PR DESCRIPTION
**Pull request by anurag12-webster**
i managed to abliterate the newer diffusiongemma, i am attaching the changed... which i did to the base herectic codebase.. with some quirks

of all the code four things needed patching to make it work..

- switched PEFT task type to FEATURE_EXTRACTION diffusiongemma (DG) doesn't implement prepare_inputs_for_generation so CAUSAL_LM was crashing.
- MoE experts are a single batched parameter [128, 2816, 704], not nn.Linear, so heretic skips them. added per-slice ablation across all 128 experts per layer
- encoder and decoder share the same weights in memory (same data_ptr) PEFT only wraps encoder so decoder never saw the LoRA delta. i have fixed it with a context manager that merges LoRA into base weights before generate() method.
- the generate() method has no output_hidden_states support so i switched it to forward hooks on encoder layers for the PCA

code is rough, happy to refactor. closes #370

### CODE REVIEW

**Comment by p-e-w** [comment]
Awesome, thanks! I'll be looking at this in detail after the 1.4 release. Perhaps we can find another diffusion model to make sure this approach generalizes.

**Comment by p-e-w** [comment]
> MoE experts are a single batched parameter [128, 2816, 704], not nn.Linear, so heretic skips them. added per-slice ablation across all 128 experts per layer

See #342 for an extended discussion of this.

/cc @rocker-zhang

**Comment by kabachuha** [comment]
What are your actual numbers: refusals / KL?

Also, did you test it with ARA/ARA-LoRA, or only with projected abliteration?

For example, [this guy's](https://huggingface.co/DuoNeural/diffusiongemma-26B-A4B-it-abliterated)'s results are not as optimistic and [this](https://huggingface.co/edwixx/diffusiongemma-26B-A4B-it-HERETIC-Uncensored/) has large KL (~0.5)

**Comment by kabachuha** [comment]
> Perhaps we can find another diffusion model to make sure this approach generalizes.

LLaDA 1.x/2.x series would be a good candidate, they are well implemented and documented

**Comment by kabachuha** [comment]
From https://huggingface.co/edwixx/diffusiongemma-26B-A4B-it-HERETIC-Uncensored

> Known issue: "own" tokens

> Because the KL divergence is 0.49, some token positions in generated outputs don't fully denoise during the diffusion process. These positions fall back to the model's mask token, which decodes as the word "own".

It's very interesting, because the "own" token spawns not only in the diffusion model version, but in the base autoregressive one as well. For example, in my overcooked LoRA training it has shown up too.

Maybe, it's some sort of a placeholder token for Gemma4 models in general?

Screenshot 2026-06-13 at 11-54-09 New Chat • Open WebUI

**Comment by anurag12-webster** [comment]
> What are your actual numbers: refusals / KL?
> 
> Also, did you test it with ARA/ARA-LoRA, or only with projected abliteration?
> 
> For example, [this guy's](https://huggingface.co/DuoNeural/diffusiongemma-26B-A4B-it-abliterated)'s results are not as optimistic and [this](https://huggingface.co/edwixx/diffusiongemma-26B-A4B-it-HERETIC-Uncensored/) has large KL (~0.5)

I have tried only with the project abliteration haven't tried with the ARA yet.

also for the numbers model currently has 13/100 refusals, KLD `0.49`, i got this at the trial 89/200.

**Comment by anurag12-webster** [comment]
> From https://huggingface.co/edwixx/diffusiongemma-26B-A4B-it-HERETIC-Uncensored
> 
> > Known issue: "own" tokens
> 
> > Because the KL divergence is 0.49, some token positions in generated outputs don't fully denoise during the diffusion process. These positions fall back to the model's mask token, which decodes as the word "own".
> 
> It's very interesting, because the "own" token spawns not only in the diffusion model version, but in the base autoregressive one as well. For example, in my overcooked LoRA training it has shown up too.
> 
> Maybe, it's some sort of a placeholder token for Gemma4 models in general?
> Screenshot 2026-06-13 at 11-54-09 New Chat • Open WebUI

Ohh i see, so my assumption abt the "own" artifact was being a denoising fallback was wrong the, and it is from gemma 4 series itself, i will update the model card with this then, thanks for confirming!

**Comment by kabachuha** [comment]
@anurag12-webster Do you have some visualizations: hidden states PCA/PaCMAP? Seeing how the refusal direction (or its equivalent) behaves in space would be incredibly helpful and useful for interpretability

**Comment by anurag12-webster** [comment]
> @anurag12-webster Do you have some visualizations: hidden states PCA/PaCMAP? Seeing how the refusal direction (or its equivalent) behaves in space would be incredibly helpful and useful for interpretability

i actually dont have it, maybe in some time i'll have it run through and shared here

**Comment by anurag12-webster** [comment]
i'll try the Llada series of models with this one and also share some visualizations along with that.
